### PR TITLE
Add a matching lerp implementation to eager mode. (#1612)

### PR DESCRIPTION
**Pull request by jjsjann123**
Fixes part of #76046

Add a matching lerp to eager mode.

Co-authored-by: jjsjann123 <alex.jann2012@gmail.com>
Co-authored-by: jjsjann123 <jiej@nvidia.com>

### CODE REVIEW

**Comment by jjsjann123** [comment]
cc'ing @kevinstephano

File: torch/csrc/jit/codegen/cuda/arith.cpp [hunk 1]
```diff
@@ -995,10 +995,28 @@ TensorView* sub_alpha(TensorView* v1, TensorView* v2, Val* v3) {
 }
 // lerp
 Val* lerp(Val* start, Val* end, Val* weight) {
+  auto casted_values =
+      promoteValues(TypePromotion::default_op_config, {start, end, weight});
+  start = casted_values[0];
+  end = casted_values[1];
+  weight = casted_values[2];
+
+  auto out_dtype =
```
Comment (ngimel) [inline torch/csrc/jit/codegen/cuda/arith.cpp:1004]: lerp in eager mode does not promote types

File: torch/csrc/jit/codegen/cuda/runtime/helpers.cu [hunk 2]
```diff
@@ -27,6 +27,44 @@ __device__ constexpr int64_t ceilDiv(int a, int64_t b) {
   return ceilDiv((int64_t)a, b);
 }
 
+// Monotonic and precise lerp is described here:
+// https://math.stackexchange.com/a/1798323
+__device__ double lerp(double start, double end, double weight) {
+  if (abs(weight) < 0.5) {
```
Comment (ngimel) [inline torch/csrc/jit/codegen/cuda/runtime/helpers.cu:33]: I don't think `abs` is correct here (and I'll need to fix it in eager too), it's needed for complex cases only.

File: torch/csrc/jit/codegen/cuda/arith.cpp (hunk 1 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/arith.cpp:1004]: promote_type just returns the higher precision type between the two. If PyTorch lerp doesn't support mixed type, this function does nothing and we'll have the same behavior from a parsed graph.

File: torch/csrc/jit/codegen/cuda/runtime/helpers.cu (hunk 2 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/runtime/helpers.cu:33]: Good catch~ Sorry for blindly following eager without double checking the actual implementation :stuck_out_tongue_closed_eyes:

File: torch/csrc/jit/codegen/cuda/arith.cpp [hunk 3]
```diff
@@ -995,10 +995,28 @@ TensorView* sub_alpha(TensorView* v1, TensorView* v2, Val* v3) {
 }
 // lerp
 Val* lerp(Val* start, Val* end, Val* weight) {
+  auto casted_values =
```
Comment (mruberry) [inline torch/csrc/jit/codegen/cuda/arith.cpp:998]: grammar nit: "cast" is the past tense of "cast"

File: torch/testing/_internal/common_methods_invocations.py [hunk 4]
```diff
@@ -13750,11 +13750,6 @@ def generate_std_var_kwargs(t: torch.Tensor, **kwargs):
            dtypesIfCUDA=floating_and_complex_types_and(torch.half, torch.bfloat16),
            dtypesIfROCM=floating_and_complex_types_and(torch.half, torch.bfloat16),
            sample_inputs_func=sample_inputs_lerp,
-           skips=(
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:13753]: Cool!

File: torch/csrc/jit/codegen/cuda/arith.cpp (hunk 3 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/arith.cpp:998]: Language is hard. :)

**Comment by jjsjann123** [comment]
I think I have addressed all issues, but I'll need a stamp to merge.

**Review by ngimel — APPROVED** [review]

**Pull request closed without merging**